### PR TITLE
[Issue #330] [WIP] Calarify the boundries between failure and errors rephrasing the latter

### PR DESCRIPTION
**Pull request by ichaida**

### CODE REVIEW

**Comment by jdegoes** [comment]
Looks good to me! Has one breaking change, `Promise.error` to `Promise.fail`, for consistency with `IO.fail`.

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 1]
```diff
@@ -443,15 +443,15 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
   /**
    * Repeats this action with the specified schedule until the schedule
    * completes, or until the first failure. In the event of failure the progress
-   * to date, together with the error, will be passed to the specified handler.
+   * to date, together with the failure value, will be passed to the specified handler.
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: With [the new terminology](https://github.com/scalaz/scalaz-zio/issues/330), "failure" represents the consequence in terms of computation that can fail in many ways. The value produced in the error channel is a "checked error". I would just make that distinction wherever it makes sense. I think it does in this case.

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 2]
```diff
@@ -309,7 +309,7 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
     new IO.Ensuring(self, finalizer)
 
   /**
-   * Executes the release action only if there was an error.
+   * Executes the release action only if there was a failure value.
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: Here is where all kinds of failures lead to cleanup, so I'd say "if this action failed".

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 3]
```diff
@@ -382,7 +382,7 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
     self.redeem[E2, A1](h, IO.now)
 
   /**
-   * Recovers from some or all of the error cases.
+   * Recovers from some or all of the failure cases.
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: checked errors

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 4]
```diff
@@ -443,15 +443,15 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
...
+   * to date, together with the failure value, will be passed to the specified handler.
    */
   final def repeatOrElse[E2, B](schedule: Schedule[A, B], orElse: (E, Option[B]) => IO[E2, B]): IO[E2, B] =
     repeatOrElse0[B, E2, B](schedule, orElse).map(_.merge)
 
   /**
    * Repeats this action with the specified schedule until the schedule
    * completes, or until the first failure. In the event of failure the progress
-   * to date, together with the error, will be passed to the specified handler.
+   * to date, together with the failure value, will be passed to the specified handler.
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: checked error

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 5]
```diff
@@ -478,15 +478,15 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
 
   /**
    * Retries with the specified schedule, until it fails, and then both the
-   * value produced by the schedule together with the last error are passed to
+   * value produced by the schedule together with the last failure are passed to
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: checked error

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 6]
```diff
@@ -478,15 +478,15 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
...
    * the recovery function.
    */
   final def retryOrElse[A2 >: A, E1 >: E, S, E2](policy: Schedule[E1, S], orElse: (E1, S) => IO[E2, A2]): IO[E2, A2] =
     retryOrElse0(policy, orElse).map(_.merge)
 
   /**
    * Retries with the specified schedule, until it fails, and then both the
-   * value produced by the schedule together with the last error are passed to
+   * value produced by the schedule together with the last failure are passed to
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: checked error

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 7]
```diff
@@ -583,7 +583,7 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
   /**
    * Runs this action in a new fiber, resuming when the fiber terminates.
    *
-   * If the fiber fails with an error it will be captured in Right side of the error Either
+   * If the fiber fails with a failure value it will be captured in Right side of the error Either
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: Here we could rephrase as such:
```
If the fiber fails, the cause will be captured in the error channel.
```
anticipating that the type of `sandbox` is going to be `IO[Cause[E], A]` very soon.

cc @jdegoes

File: core/shared/src/main/scala/scalaz/zio/Promise.scala [hunk 8]
```diff
@@ -69,10 +69,10 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
   final def complete(a: A): IO[Nothing, Boolean] = done(ExitResult.Completed[E, A](a))
 
   /**
-   * Fails the promise with the specified error, which will be propagated to all
+   * Fails the promise with the specified a failure value, which will be propagated to all
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/Promise.scala]: checked error

File: core/shared/src/main/scala/scalaz/zio/IO.scala (hunk 7 above)
Comment (jdegoes) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: 👍

**Comment by jdegoes** [comment]
@ichaida This pull request needs a few updates to reconcile conflicts and modify terminology a bit as per @regiskuckaertz suggestions. I will close this PR, so no rush, but if you have time to fix the conflicts and update the terminology, we're ready to go on this one!

**Pull request closed without merging**